### ISSUE TITLE
LDAP username and validation. 

### ISSUE THREAD

**Issue by k-apc** [AxisCommunications/goalie-url-shortener#44]
hi there, 

I'm trying to set it up to use our ldap server and I have a couple of questions/issues

1.- we use usernames (not emails) to authenticate, therefore I'm leaving the `LDAP_EMAIL_DOMAIN=` field empty in the `.env` file, but it's still adding the `@` and that would never work. I seeing this  in the ldap server logs, so it's being able to reach it.

```
May 10 22:55:56 ldaps slapd[482]: conn=25163 fd=27 ACCEPT from IP=xxx.xxx.xxx.xxx:39106 (IP=xxx.xxx.xxxx.xxx:636)
May 10 22:55:56 ldaps slapd[482]: conn=25163 fd=27 TLS established tls_ssf=256 ssf=256
May 10 22:55:56 ldaps slapd[482]: conn=25163 op=0 do_bind: invalid dn (XXXXXXX@)
May 10 22:55:56 ldaps slapd[482]: conn=25163 op=0 RESULT tag=97 err=34 text=invalid DN
```

2.- we use an admin user to run the ldap validation, and every single ldap capable platform I've used has "admin username" and "password" fields (sometime with different names) but it looks like there's none in this one. Again... hints?

Thanks in advance

**Comment by AntonFriberg** [comment]
Hi! Thanks for reaching out! I have needed some external input in order to motivate me to spend more time on this repository. I will have a quick refresher of how the authentication works since it was a while ago that I implemented it.

I know that we decided to avoid a binding user (or an LDAP admin if you prefer) since it only need basic authentication against the server and not really any lookups or searches. 

I will get back to you with more information tomorrow hopefully!

**Comment by k-apc** [comment]
hey! thanks for taking a look! just for you to know, we are not using a self signed tls/ssl cert (but a valid letsencrypt created one), so I think we don't need to import it, correct?

looking forward!

**Comment by k-apc** [comment]
hi @AntonFriberg, any update on this?
Thanks!

**Comment by AntonFriberg** [comment]
Hi again!

Sorry for the late reply. Was not able to get around to it until today. I noticed that we have a few issues with old packages that needs some attention but you might have solved that yourself. Regarding your questions.

1.  I see the problem with the `LDAP_EMAIL_DOMAIN` solution currently in place. I upload a fix that removes the `@` symbol if the field is empty.
2. It is important to consider that this project was created during a summer by then *student workers* with very little AD experience. I have since learnt a lot more about how the login process is usually implemented. You are correct that most AD implementations make sure to keep the AD lookups fast by having a bound connection that is authenticated with a "admin" account or sometimes called "bind" user. In this project, that is not the case. Instead we re-establish a new connection for each log-in. This might seem like a slow approach but in our testing we have not seen the need to change how the implementation is done and response times are usually below 20 ms.
  We found that the connection initialization was always fast, but the responses by the AD server was very slow on initial requests with new users (which was all logins in our case). This was fixed by specifying `get_info=None` in https://github.com/AxisCommunications/goalie-url-shortener/blob/master/backend/api/auth.py#L55 . The reason why this configuration speeds up the lookups significantly, reducing the need for a bind user, is that it configures the server connection to no longer return the AD schema on reads. https://ldap3.readthedocs.io/en/latest/server.html . We do not need this information since we only check for success on authentication (login success) and utilize the fact that we should have search results on our search if the user is an Admin i.e the authorization (user found in AD group). 

I am going to get to work on implementing your first request. Let me know if you have any other questions.
